### DISCUSSION TITLE
why add assert in `const_reference operator[](const typename object_t::key_type& key) const` instead of just throw a exception

### DISCUSSION THREAD

**Discussion by h20282** [nlohmann/json#4376] [Q&A] [answered]
I noticed there is a difference between adding or not adding 'const' when using json::operator []

```c++
#include <iostream>

#include "nlohmann/json.hpp"

int main() {
    nlohmann::json json{{"b", "1"}}; // without const
    try {
        int a = json["a"]; // crash
        std::cout << "a = " << a << std::endl;
    } catch (const nlohmann::json::exception &e) {
        std::cout << e.what() << std::endl;
    }

    std::cout <<"end";
}
```

```C++
#include <iostream>

#include "nlohmann/json.hpp"

int main() {
    const nlohmann::json json{{"b", "1"}}; // with const
    try {
        int a = json["a"];
        std::cout << "a = " << a << std::endl;
    } catch (const nlohmann::json::exception &e) {
        std::cout << e.what() << std::endl; // exception catched
    }

    std::cout <<"end";
}
```

To figure out why they are different, I got this code:
![image](https://github.com/nlohmann/json/assets/43230320/568f8ef7-8a5d-4cf8-ab1f-62c29648cd30)

https://github.com/nlohmann/json/blob/develop/include/nlohmann/json.hpp#L2147

so, why not just throw a exception?

**Comment by nlohmann** [comment, marked as answer]
See https://json.nlohmann.me/features/element_access/unchecked_access/#overview

Basically, `operator[]` offers unchecked access. Use `at` if you prefer exceptions instead of undefined behavior.
